### PR TITLE
Fix GC sync race condition

### PR DESCRIPTION
**Pull request by ironcladlou**
Remove faulty diff detection logic from GC sync which leads to a race
condition: If the GC's discovery client is returning a fully up to date
view of server resources during the very first GC sync, the sync
function will never sync monitors or reset the REST mapper unless
discovery changes again. This causes REST mapping to fail for any custom
types already present in discovery.

Fixes https://github.com/kubernetes/kubernetes/issues/56262.

```release-note
NONE
```

/cc @liggitt @caesarxuchao

### CODE REVIEW

**Comment by ironcladlou** [comment]
So far no flakes over 300 `stress` runs.

**Comment by ironcladlou** [comment]
Bug was introduced in 3d6d57a18f12c575de77056be4316954a3e0ed20

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 1]
```diff
@@ -214,7 +207,13 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.DiscoveryInterface, p
 		}
 		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
 			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
+			return
 		}
+
+		// Finally, keep track of our new state. Do this after all preceding steps
+		// have succeeded to ensure we'll retry on subsequent syncs if an error
+		// occured.
+		oldResources = newResources
```
Comment (ironcladlou) [inline pkg/controller/garbagecollector/garbagecollector.go]: Although it hasn't been reported anywhere (that I know of), @liggitt noticed this potential bug during the course of reviewing the original patch. If there are no objections, I'd like to bundle the fix in this PR.

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 2]
```diff
@@ -208,7 +207,13 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.DiscoveryInterface, p
 		}
 		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
 			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
+			return
 		}
+
+		// Finally, keep track of our new state. Do this after all preceding steps
```
Comment (liggitt) [inline pkg/controller/garbagecollector/garbagecollector.go]: add a V(2) log message here that sync completed... want to be able to pair with the log message from line 183 to know resync completed

Comment (ironcladlou) [reply pkg/controller/garbagecollector/garbagecollector.go]: Done

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 3]
```diff
@@ -212,9 +205,19 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.DiscoveryInterface, p
 			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
 			return
 		}
+		// TODO: WaitForCacheSync can block forever during normal operation. Could
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/garbagecollector.go]: When will it block forever?

**Comment by caesarxuchao** [comment]
The patch lgtm. I asked a question regarding the todo.

File: pkg/controller/garbagecollector/garbagecollector.go (hunk 3 above)
Comment (liggitt) [reply pkg/controller/garbagecollector/garbagecollector.go]: if newResources contains GVKs that are being removed, and are gone by the time we get here, I think blocks until the informers for newResources are all synced, which will never happen successfully for the now-missing GVKs.

1. add new resource (CRD, add-on APIService, etc)
2. GC GetDeletableResources sees new resource, notices something has changed
3. delete new resource
4. GC resyncMonitors sets up new monitors/informers for newResources, including for the now-gone resource
5. GC WaitForCacheSync waits until the informers are all synced, which will never succeed for the now-gone resource

**Comment by spiffxp** [comment]
/status approved-for-milestone
getting rid of the flakiest PR flake? sure

**Comment by ironcladlou** [comment]
>You must be a member of the kubernetes/kubernetes-milestone-maintainers github team to add status labels.
😛

File: pkg/controller/garbagecollector/garbagecollector.go (hunk 3 above)
Comment (caesarxuchao) [reply pkg/controller/garbagecollector/garbagecollector.go]: Shall we plumb a flag into the reflector to instruct it to not retry listing if the error is 404?

Comment (liggitt) [reply pkg/controller/garbagecollector/garbagecollector.go]: I don't know... requires more thought. @ironcladlou can you open an issue to track this to make sure it stays high on our radar?

Comment (jennybuckley) [reply pkg/controller/garbagecollector/garbagecollector.go]: Is there an open issue for this? Based on the logs, I think it is the cause of #60037

Comment (jennybuckley) [reply pkg/controller/garbagecollector/garbagecollector.go]: @liggitt @ironcladlou I couldn't find the issue tracking this race

I am wondering if something like https://github.com/kubernetes/kubernetes/pull/61057 will be enough to fix the race?

Comment (ironcladlou) [reply pkg/controller/garbagecollector/garbagecollector.go]: @jennybuckley looks like I neglected to create the followup issue for this. I've been slammed the past couple weeks with things unrelated to Kube but I'll try to take a look at https://github.com/kubernetes/kubernetes/pull/61057 today. Thank you!

**Comment by jennybuckley** [comment]
I just noticed that the bug this fixes is also in 1.8, would it be useful to cherrypick this back to 1.8?

**Pull request merged**